frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,786 @@
+					}
+					extra = true;
+				},
+				BenchmarkAttrKeyword::SkipMeta => {
+					if skip_meta {
+						return Err(content.error("`skip_meta` can only be specified once"))
+					}
+					skip_meta = true;
+				},
+			}
+		}
+		Ok(BenchmarkAttrs { extra, skip_meta })
+	}
+}
+
+/// Represents the parsed extrinsic call for a benchmark
+#[derive(Clone)]
+enum BenchmarkCallDef {
+	ExtrinsicCall { origin: Expr, expr_call: ExprCall }, // #[extrinsic_call]
+	Block(ExprBlock),                                    // #[block]
+}
+
+/// Represents a parsed `#[benchmark]` or `#[instance_banchmark]` item.
+#[derive(Clone)]
+struct BenchmarkDef {
+	params: Vec<ParamDef>,
+	setup_stmts: Vec<Stmt>,
+	call_def: BenchmarkCallDef,
+	verify_stmts: Vec<Stmt>,
+	extra: bool,
+	skip_meta: bool,
+}
+
+impl BenchmarkDef {
+	/// Constructs a [`BenchmarkDef`] by traversing an existing [`ItemFn`] node.
+	pub fn from(item_fn: &ItemFn, extra: bool, skip_meta: bool) -> Result<BenchmarkDef> {
+		let mut params: Vec<ParamDef> = Vec::new();
+
+		// parse params such as "x: Linear<0, 1>"
+		for arg in &item_fn.sig.inputs {
+			let span = arg.span();
+			let invalid_param = || {
+				return Err(Error::new(span, "Invalid benchmark function param. A valid example would be `x: Linear<5, 10>`.", ))
+			};
+
+			let FnArg::Typed(arg) = arg else { return invalid_param() };
+			let Pat::Ident(ident) = &*arg.pat else { return invalid_param() };
+			let name = ident.ident.to_token_stream().to_string();
+			let typ = &*arg.ty;
+			let Type::Path(tpath) = typ else { return invalid_param() };
+			let Some(segment) = tpath.path.segments.last() else { return invalid_param() };
+			let args = segment.arguments.to_token_stream().into();
+			let Ok(args) = syn::parse::<RangeArgs>(args) else { return invalid_param() };
+			let Ok(start) = args.start.base10_parse::<u32>() else { return invalid_param() };
+			let Ok(end) = args.end.base10_parse::<u32>() else { return invalid_param() };
+
+			params.push(ParamDef { name, typ: typ.clone(), start, end });
+		}
+
+		// #[extrinsic_call] / #[block] handling
+		let Some(Ok((i, call_def))) = (&item_fn.block.stmts).iter().enumerate().find_map(|(i, child)| {
+			match child {
+				Stmt::Semi(Expr::Call(expr_call), _semi) => { // #[extrinsic_call] case
+					(&expr_call.attrs).iter().enumerate().find_map(|(k, attr)| {
+						let Some(segment) = attr.path.segments.last() else { return None; };
+						let Ok(_) = syn::parse::<keywords::extrinsic_call>(segment.ident.to_token_stream().into()) else { return None; };
+						let mut expr_call = expr_call.clone();
+
+						// consume #[extrinsic_call] tokens
+						expr_call.attrs.remove(k);
+
+						// extract origin from expr_call
+						let origin = match expr_call.args.first() {
+							Some(arg) => arg.clone(),
+							None => return Some(Err(Error::new(expr_call.args.span(), "Single-item extrinsic calls must specify their origin as the first argument."))),
```

Comment:
This error will unfortunately get eaten up by the else block down there. Perhaps let-else is actually not a good use here, because you actually want to return this error.

Reply:
Restructured as a match at the top level and now it shouldn't be suppressed